### ISSUE TITLE
Feature request: Automatically set button state colors

### ISSUE THREAD

**Issue by ethanroday** [OfficeDev/office-ui-fabric-react#7585]
**EDIT**: I should clarify that these screenshots are from the Fluent UI experiment, but the overall question still stands.

-------------------

Sometimes you need a one-off button that has a color outside of your main theme.
It would be great if there was an easy way to set the color for the default button state and have all of the other colors set automatically. As such, if I have the following code, here's what I see:

```tsx
<DefaultButton className={[ColorClassNames.redBackground, ColorClassNames.white].join(" ")}>
  Don't Click This!
</DefaultButton>
```

Default:
![image](https://user-images.githubusercontent.com/896670/50923462-e326bb80-1401-11e9-92c5-33de2b32759a.png)

Hover:
![image](https://user-images.githubusercontent.com/896670/50923493-fb96d600-1401-11e9-8b97-570f08de1def.png)

Pressed:
![image](https://user-images.githubusercontent.com/896670/50923529-123d2d00-1402-11e9-8618-a1c3d9ecd0b9.png)

---------------

I can partially solve this by using some additional members from `ColorClassNames`, but the pressed state is still a little weird, and it would be great not to have to have this much boilerplate:

```tsx
<DefaultButton
  className={[
    ColorClassNames.redBackground,
    ColorClassNames.redBorder,
    ColorClassNames.redDarkBackgroundHover,
    ColorClassNames.redDarkBorderHover,
    ColorClassNames.white,
    ColorClassNames.whiteHover
  ].join(" ")}
>
  Don't Click This!
</DefaultButton>
```

Default:
![image](https://user-images.githubusercontent.com/896670/50923879-d5be0100-1402-11e9-9f4a-23b61b10d57a.png)

Hover:
![image](https://user-images.githubusercontent.com/896670/50923859-ce96f300-1402-11e9-964e-c83ba64bea0d.png)

Pressed:
![image](https://user-images.githubusercontent.com/896670/50923828-c048d700-1402-11e9-92a0-b6d6220fcb1f.png)

**Comment by ethanroday** [comment]
Fluent UI as in: https://fluentfabric.azurewebsites.net/#/overview
`ColorClassNames` is a set of class names that I found sort of by accident. The classes define text color, background color, and border color on both default and hover states.

**Comment by JasonGore** [comment]
I'm seeing a couple of different asks here:
1. Components should be able to take in one color and derive the others. 
    * For example, hover should be equal to a slightly darker shade of default. (At least that's how I interpret this request.)
    * This doesn't have anything to do with Slots. It's basically saying components should develop a relation between colors that isn't in place now. However, I'm not sure if it will always hold as you can't easily say that hover is a slightly darker version of default, because then what do we do for inverted scenarios?
2. Easy access to button tokens state to determine coloring from a consumer level.
    * My Tokens PR (#7552) shows how Tokens can be used to affect button coloring based on tokens and state. See `TeamsCustomizations.ts` to see how this could be done using tokens and button state.

**Comment by Vitalius1** [comment]
@ethanroday I completely forgot about the `@uifabric/styling` package you are using in this example. I believe it is meant to port in TS everything Fabric Core is having in SASS.

As Jason said, it's somewhat incorrect to expect from specifying `redBackground` for the default state to automatically set a color for the hover state. What if someone decides to have for the hover state `redBackgroundHover` and someone else `redDarkBackgroundHover` (how you did above). Someone can decide to go nuts and have `purpleBackgrounHover`. I think these colors are meant to give you somewhat semantic meaning to a color and it's each consumer's choice on how to combine them.

As for the pressed state (and also focused), it's something that I think should be discussed with designers and added to the properties of `ColorClassNames` for each color. Today we only have these states for each color:
https://github.com/OfficeDev/office-ui-fabric-react/blob/04558fe11e54e6c0f1db85e14c32f65cfb1d502f/packages/styling/src/classNames/ColorClassNames.ts#L192-L197

But these should be added along with updating Fabric Core also. @mikewheaton do you know why we only have color classNames for background and border in default and hover states only. Where there any plans to add similar colors for focus and active states?

**Comment by mikewheaton** [comment]
> @mikewheaton do you know why we only have color classNames for background and border in default and hover states only. Where there any plans to add similar colors for focus and active states?

I believe these were added to be equivalent to [Fabric Core's color classes](https://github.com/OfficeDev/office-ui-fabric-core/blob/master/src/sass/_Color.scss), which target only `color` and `background-color` (with a hover variant for each).

**Comment by Vitalius1** [comment]
Were there any plans or discussions on whether add or not add other states (focus/active)? It also has covered the `border-color`.

**Comment by mikewheaton** [comment]
No, there weren't any plans for focus/active classes.

**Comment by Vitalius1** [comment]
@dzearing @phkuo as package owners... is this something we would want to add? (meaning some additional states (focus and active) for each `ColorClassName`). I think this can quickly get out of control if taking into consideration how many variations should be added. At the same time, this is the equivalent of the ones we have in Fabric Core and it would be expected to add them there too for making them available in SASS.

**Comment by phkuo** [comment]
Focus state tends to be just an outline, or an outline in combination with the control's hoverstate, so a separate focus set of classes isn't necessary. Active is a state that we do want to support.
I don't remember where ColorClassName came from, you'll have to review that with the Fabric team about how and where it's meant to be used.

**Comment by dzearing** [comment]
I don't think `ColorClassNames` is what we want here. Nothing at least in Fabric uses it, and as we've moved on to css in js and theming, it has become far less important. Tokens is how we'd answer this.

```ts
// Use theme and props if needed.
const redButtonTokens = ({ theme }) => {
 background: 'red',
 hoveredBackground: 'darkRed',
 pressedBorder: '1px solid darkRed'
};

<Button tokens={ redButtonTokens } text='I am a button' />

// or, make it reusable:

const RedButton = props => <Button { ...props } tokens={ redButtonTokens } />;

<RedButton text='I am a red button' />
```

I don't think we *need* to support calculated token values here to support the scenario, as you could generate the classnames in your token calculation using helpers if needed.

**Comment by khmakoto** [comment]
@ethanroday Just to be completely sure here, can you clarify what is the expected behavior on the pressed button? Is it a red background color with a dark red border color?

**Comment by Vitalius1** [comment]
@khmakoto I think it's not as much the color being the ask here as more the better alternative to easily style a component (button here just being an example). @ethanroday was using `ColorClassNames` from the `styling` package and was wondering if there could be an easier way with less boilerplate. Also, there were some issues with the `active` state color not being present in the `ColorClassNames` object. What @dzearing meant is that we would not want to invest in adding those extra classNames and instead try using the new tokens pattern. But because this pattern is limited to only the components in the experiments package I don't think it would be of any use to the issue author. For demo purposes though, it would be nice to see a codepen with tokens in usage 😄 .

@ethanroday I talked to @dzearing offline and it appears we would not want to add any more color classNames to the `ColorClassNames` object as it would bloat unnecessary the styling package (especially that this was added to just bring everything that Fabric Core had in SASS). For reference what `Tokens` are and how we plan to improve styling a component can be found here #7552 

cc: @JasonGore if he wants to add anything else.
